tests/MongoDB.Driver.Core.Tests/Core/Bindings/CoreSessionTests.cs

```diff
@@ -241,17 +248,256 @@ public void WasUsed_should_call_serverSession()
+
+            var e = exception.Should().BeOfType<NotSupportedException>().Subject;
+            e.Message.Should().Be("StartTransaction cannot determine if transactions are supported because there are no connected servers.");
+        }
+
+        [Theory]
+        [InlineData("PN", "Server version 3.6.99 does not support the Transactions feature.")]
+        [InlineData("PN,ST", "Server version 3.6.99 does not support the Transactions feature.")]
+        [InlineData("PT,SN", "Server version 3.6.99 does not support the Transactions feature.")]
+        [InlineData("RN", "Server version 3.6.99 does not support the Transactions feature.")]
+        [InlineData("RN,RT", "Server version 3.6.99 does not support the Transactions feature.")]
+        [InlineData("RT,RN", "This version of the driver does not support sharded transactions.")]
+        public void EnsureTransactionsAreSupported_should_throw_when_any_connected_data_bearing_server_does_not_support_transactions(string scenarios, string expectedMessage)
+        {
+            var clusterId = new ClusterId(1);
+            var servers =
+                SplitScenarios(scenarios)
+                .Select((scenario, i) =>
+                {
+                    var endPoint = new DnsEndPoint("localhost", 27017 + i);
+                    var serverId = new ServerId(clusterId, endPoint);
+                    var type = MapServerTypeCode(scenario[0]);
+                    var supportsTransactions = MapSupportsTransactionsCode(scenario[1]);
+                    var version = supportsTransactions ? (type == ServerType.ShardRouter ? new SemanticVersion(4, 2, 0) : Feature.Transactions.FirstSupportedVersion) : Feature.Transactions.LastNotSupportedVersion;
+                    return CreateServerDescription(serverId, endPoint, ServerState.Connected, type, version);
+                })
+                .ToList();
+            var cluster = CreateClusterDescription(clusterId, servers: servers);
+            var subject = CreateSubject(cluster);
+
+            var exception = Record.Exception(() => subject.EnsureTransactionsAreSupported());
+
+            var e = exception.Should().BeOfType<NotSupportedException>().Subject;
+            e.Message.Should().Contain(expectedMessage);
+        }
+
+        [Theory]
+        [InlineData("3.6", ServerType.ReplicaSetPrimary, "Server version 3.6.0 does not support the Transactions feature.")]
+        [InlineData("4.0", ServerType.ReplicaSetPrimary, null)]
+        [InlineData("4.2", ServerType.ReplicaSetPrimary, null)]
+        [InlineData("3.6", ServerType.ShardRouter, "Server version 3.6.0 does not support the Transactions feature.")]
+        [InlineData("4.0", ServerType.ShardRouter, "Server version 4.0.0 does not support the ShardedTransactions feature.")]
+        [InlineData("4.2", ServerType.ShardRouter, "This version of the driver does not support sharded transactions.")]
```

Comment:
These tests duplicate some earlier tests but make it easier to verify what the exception message (if any) is for each server version for either a replica set or sharded cluster.